lib/compress/index.js

```diff
@@ -1598,6 +1598,53 @@ def_optimize(AST_Switch, function(self, compressor) {
     while (i < len) eliminate_branch(self.body[i++], body[body.length - 1]);
     self.body = body;
 
+    let default_or_exact = default_branch || exact_match;
+
+    // group equivalent branches so they will be located next to each other,
+    // that way the next micro-optimization will merge them.
+    // ** bail micro-optimization if not a simple switch case with breaks
+    if (body.every((branch, i) =>
+        (branch === default_or_exact || !branch.expression.has_side_effects(compressor))
+        && (branch.body.length === 0 || branch.body.some(aborts) || body.length - 1 === i))
+    ) {
+        for (let i = 0; i < body.length; i++) {
+            const branch = body[i];
+            if (branch.body.length === 0) continue;
+            if (!aborts(branch)) continue;
+            for (let j = i + 1; j < body.length; j++) {
+                const next = body[j];
+                if (next.body.length === 0) continue;
+                const last_branch = j === (body.length - 1);
+                const equivalentBranch = branches_equivalent(next, branch, false);
+                if (equivalentBranch || (last_branch && branches_equivalent(next, branch, true))
+                ) {
+                    if (!equivalentBranch && last_branch) {
+                        if (!is_break(next.body[next.body.length - 1])) {
+                            next.body.push(make_node(AST_Break));
+                        }
+                    }
+
+                    // let's find previous siblings with empty fallthrough...
+                    let x = j - 1;
+                    let fallthroughDepth = 0;
+                    while (x > i) {
+                        let previousSiblingBranch = body[x--];
+                        if (previousSiblingBranch.body.length === 0) {
```

Comment:
We can handle side-effectless branches, too.

```suggestion
                        if (is_inert_body(previousSiblingBranch)) {
```

Reply:
@jridgewell not sure what you mean by that, and what does `inert_body` mean?

Reply:
`inert_body` means the body doesn't effect the runtime evaluation of the switch. Eg,

```js
switch (id(3)) {
  case 0:
    console.log('PASS');
    break;
  case 1:
    console.log("not inert because I cause side-effect");
    break;
  case 2:
    "I am an inert branch even though I'm not empty";
  case 3:
    console.log('PASS');
    break;
}
```

This could output (if we ignore the later passes):

```js
switch (id(3)) {
  case 0:
    console.log('PASS');
    break;
  case 2:
    "I am an inert branch even though I'm not empty";
  case 3:
    console.log('PASS');
    break;
  case 1:
    console.log("not inert because I cause side-effect");
    break;
}
```

We can make this change in a follow-up.

Reply:
Is there anyway these inert branches can be simply eliminated or emptied an early step? This would reduce the amount of checks for them.

Reply:
(in a follow-up)

Reply:
Oh no. I see this is done here

https://github.com/terser/terser/pull/1070/files#diff-db998527f1e5c1b3ca86a50c09e588393ef935ecc0d8de75952127344ebf26efL1657

So it may be simply a matter of ordering.

Reply:
@jridgewell thanks for the explanation, I updated this code.
@fabiosantoscode I'm not sure I want to do any reordering, as it might be a bit out-of-scope for this optimization, and I really like the fact that this optimization is nicely encapsulated. LMKYT

Reply:
Yeah, I've read it top to bottom again and it is pretty self-contained. Pretty good in a project that relies on a lot of mutation :D